elements/pl-barcode-scan/pl-barcode-scan.py

```diff
@@ -0,0 +1,82 @@
+import prairielearn as pl
+import lxml.html
+import chevron
+
+BARCODE = '_pdf_barcode_scan'
+BLANK_ANSWER = ''
+REQUIRED = 'required'
+REQUIRED_DEFAULT = True
+
+
+def crc16arc(msg):
+    """
+    CRC-16/ARC
+    This algorithm has to match the 'js-crc' npm algorithm, which was assumed to be the same as this.
+    https://stackoverflow.com/questions/63167168/trying-to-find-the-input-for-crc16-ibm
+    """
+    crc = 0
+    for b in msg:
+        crc ^= b
+        for _ in range(8):
+            crc = (crc >> 1) ^ 0xa001 if crc & 1 else crc >> 1
+    return crc
+
+
+def is_invalid_barcode(barcode):
+    barcode = barcode.lower()
+    sha16 = barcode[len(barcode) - 4:len(barcode)]
+    base32 = barcode.replace(sha16, '')
+    recomputed_sha16 = hex(crc16arc(bytes(base32, encoding='utf-8')))
+    if sha16 in recomputed_sha16:
+        return False
+    else:
+        return True
+
+
+def prepare(element_html, data):
+    lxml.html.fragment_fromstring(element_html)
+
+
+def render(element_html, data):
+    element = lxml.html.fragment_fromstring(element_html)
+    required = pl.get_boolean_attrib(element, REQUIRED, REQUIRED_DEFAULT)
+    uuid = pl.get_uuid()
+
+    if data['panel'] == 'question':
+        html_params = {
+            'question': True,
+            'uuid': uuid,
+            'required': required
+        }
+    else:
+        html_params = {
+            'uuid': uuid,
+            'question': False,
+            'parse-error': data['format_errors'].get(BARCODE, None),
+            'barcode': data['submitted_answers'].get(BARCODE)
+        }
+
+    with open('pl-barcode-scan.mustache', 'r', encoding='utf-8') as f:
+        html = chevron.render(f, html_params).strip()
+    return html
+
+
+def parse(element_html, data):
+    element = lxml.html.fragment_fromstring(element_html)
+    submitted_barcode = data['submitted_answers'].get(BARCODE, '').strip()
+    required = pl.get_boolean_attrib(element, REQUIRED, REQUIRED_DEFAULT)
+
+    # TO DO: Figure out logic for if we need more barcodes per instance question submission.
+    # ie. Student submits two pieces of paper to be scanned.
+    del data['submitted_answers'][BARCODE]
```

Comment:
I think it is a good idea. I'm hesitant to delve into this now, but I will see what I can do under two hours today. I really need to start looking at the next task. Please let me know how strongly you feel about this change right now.

Reply:
My concern is that if we *don't* design this for multiple barcodes, we'll then need to maintain two logic paths every time we deal with barcodes (one for the "single string" case, one for the "array of strings" case) to deal with the "legacy" data that this will create. Even if we only support students entering a single barcode for now, ensuring that we store it as an array of `["barcode"]` and that we can correctly index + access that will save us trouble in the future. This is a "strong convictions, loosely held" thing - I'd be happy to be convinced that it won't be difficult to make this change after the fact in the future!

Reply:
Just do a migration on the index to convert it to an array before the release. The index should make it easy to discover which submissions need to be updated. I don't have much experience with Postgres migrations, but it looks like your framework makes it fairly straight forward.

The conversation specified one barcode per submission. There wasn't a push to move towards multiple barcodes per submission just yet.